src/types.rs

```diff
@@ -166,34 +219,67 @@ impl TypeDefinition {
                 BuiltInTypeDefinition::Atomic(ion_type, is_nullable) => {
-                    if *is_nullable == Nullability::NotNullable && value.is_null() {
-                        return Err(Violation::new(
-                            "type_constraint",
-                            ViolationCode::InvalidNull,
-                            &format!("expected type {:?} doesn't allow null", ion_type),
-                        ));
-                    }
-                    if value.ion_type() != *ion_type {
-                        return Err(Violation::new(
-                            "type_constraint",
-                            ViolationCode::TypeMismatched,
-                            &format!("expected type {:?}, found {:?}", ion_type, value.ion_type()),
-                        ));
+                    // atomic types doesn't include document type
+                    match value {
+                        IonSchemaElement::SingleElement(element) => {
+                            if *is_nullable == Nullability::NotNullable && element.is_null() {
+                                return Err(Violation::new(
+                                    "type_constraint",
+                                    ViolationCode::InvalidNull,
+                                    &format!("expected type {:?} doesn't allow null", ion_type),
+                                ));
+                            }
+                            if element.ion_type() != *ion_type {
+                                return Err(Violation::new(
+                                    "type_constraint",
+                                    ViolationCode::TypeMismatched,
+                                    &format!(
+                                        "expected type {:?}, found {:?}",
+                                        ion_type,
+                                        element.ion_type()
+                                    ),
+                                ));
+                            }
+
+                            Ok(())
+                        }
+                        IonSchemaElement::Document(document) => {
+                            return Err(Violation::new(
+                                "type_constraint",
+                                ViolationCode::TypeMismatched,
+                                &format!("expected type {:?}, found document", ion_type),
+                            ));
+                        }
                     }
-                    Ok(())
                 }
                 BuiltInTypeDefinition::Derived(other_type) => {
+                    if other_type.name() == &Some("document".to_owned()) {
+                        // Verify whether the given derived type is document
+                        // And check if it is using enum variant IonSchemaElement::Document
+                        if !matches!(value, IonSchemaElement::Document(_)) {
```

Comment:
Might be nice to have `is_document()` and `is_element()` convenience methods on the `IonSchemaElement` type.

Reply:
I already had `as_element()` defined as:
```rust
pub fn as_element(&self) -> Option<&OwnedElement> {
        match self {
            IonSchemaElement::SingleElement(element) => Some(element),
            IonSchemaElement::Document(_) => None,
        }
    }
```
I have similarly created `as_docuemnt()` and used it for type check as below:
```rust
if value.as_document() == None {
     // error if not a document
}
```